### ISSUE TITLE
`instanceMatch`

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#8460]
Say you are writing a partial function to retrieve elements based on keys in a map. When it fails you want to print the key. You don't want to require a `ToString` instance but if it is there you want to use it! We will have this information during monomorphization, so why not allow it?

```
def getCrash(key: k, m: Map[k ,v]): v \ KeyNotFound = match Map.get(key, m) {
    instanceMatch {
        case ToString[k] => do KeyNotFound.keyNotFound("key not found: ${key}")
        case _ => do KeyNotFound.keyNotFound("key not found")
    }
}
```

**Comment by JonathanStarup** [comment]
@mlutze what do you think? how painful would it be to add?

**Comment by mlutze** [comment]
> @mlutze what do you think? how painful would it be to add in typing? 

The local constraints scare me...

**Comment by JonathanStarup** [comment]
yeah me too. could we bind it to a local type variable somehow to make the trait itself not local? or some other trick?

**Comment by magnus-madsen** [comment]
An instance may only exist if some other instance exists. So this makes it even harder, no?

**Comment by JonathanStarup** [comment]
maybe, but we have the ground type for k so i should still be okay
